packages/mds-policy-service/service/validators.ts

```diff
@@ -104,74 +86,128 @@ const checkRateFieldsIfConditionSchema = () => ({
   },
   then: {
     properties: {
-      currency: stringSchema()
+      currency: { type: 'string' }
     },
     required: ['currency']
   }
-})
+}
+
+type MICRO_MOBILITY_VEHICLE_EVENT_ARRAY_ENUM = {
+  type: 'array'
+  items: { type: 'string'; enum: MICRO_MOBILITY_VEHICLE_EVENT[] | [] }
+  nullable: true
+}
+
+const micromobilityStateMap = MICRO_MOBILITY_VEHICLE_STATES.reduce<{
+  [k in MICRO_MOBILITY_VEHICLE_STATE]?: MICRO_MOBILITY_VEHICLE_EVENT_ARRAY_ENUM
+}>((acc, state) => {
+  acc[state] = {
+    type: 'array',
+    items: enumSchema([...MICRO_MOBILITY_VEHICLE_EVENTS]),
+    nullable: true
+  }
+  return acc
+}, {}) as { [S in MICRO_MOBILITY_VEHICLE_STATE]: MICRO_MOBILITY_VEHICLE_EVENT_ARRAY_ENUM }
```

Comment:
Is this `as` necessary now that you've got the type being passed in as a generic to the reduce? And if it is required, is it possible to move it further in, onto the `{}` initialization for the accumulator?

Reply:
oh good, this re-declaration from key-optional to key-required was needed during early phases of the validator development. Thankfully, not anymore!